contracts/gateway/EIP20Gateway.sol

```diff
@@ -477,7 +485,42 @@ contract EIP20Gateway is GatewayBase {
             bytes32(0),
             true
         );
+    }
+
+    /**
+     * @notice Return the revert penalty to the staker. Only valid for
+     *         a message transition from DeclaredRevocation -> Progressed.
+     *
+     * @dev Should only be called from progressStakeWithProof. This function
+     *      exists to avoid a stack too deep error.
+     *
+     * @param _messageHash Message hash.
+     * @param _staker Staker address.
+     * @param _oldMessageStatus Message status before progressing.
+     * @param _newMessageStatus Message status after progressing.
+     */
+    function tryReturnPenaltyToStaker(
+        bytes32 _messageHash,
+        address _staker,
+        MessageBus.MessageStatus _oldMessageStatus,
+        MessageBus.MessageStatus _newMessageStatus
+    )
+      private
+    {
+        if (_oldMessageStatus != MessageBus.MessageStatus.DeclaredRevocation) {
+            return;
+        }
+        if (_newMessageStatus != MessageBus.MessageStatus.Progressed) {
```

Comment:
As it is called from `progressStakeWithProof` I think only checking for `MessageBus.MessageStatus.DeclaredRevocation` will be sufficient here.
We may not need `_newMessageStatus` params.

Reply:
Or may be just the following code in `progressStakeWithProof`
```
if(messageBox.outbox[_messageHash] == MessageBus.MessageStatus.DeclaredRevocation) {
	uint256 penalty = penaltyFromBounty(stakes[_messageHash].bounty);
	// transfer the penalty amount
        require(baseToken.transfer(_staker, penalty));
}
```

Reply:
I tried removing those, but then the `should fail when the message status in source is declared revocation and in the target is declared` test fails with `Message on target must be Progressed.`, so it doesn't seem that the other checks are sufficient?

Reply:
We need `_newMessageStatus`(Inbox message status). It should return the penalty only if the inbox is `progressed`. If it's declared on target and declare revocation on source then possible next step is confirmRevocation on target.  

Penalty should be returned only if target is progressed.